### PR TITLE
feat: increase precision for important calculations

### PR DESCRIPTION
**Pull request by steffenix**
fixes: #221

### CODE REVIEW

File: contracts/Vault.vy [hunk 1]
```diff
@@ -899,7 +905,7 @@ def _sharesForAmount(amount: uint256) -> uint256:
     # See dev note on `deposit`.
     if self._totalAssets() > 0:
         # NOTE: if sqrt(token.totalSupply()) > 1e37, this could potentially revert
-        return ((10 ** 3 * (amount * self.totalSupply)) / self._totalAssets()) / 10 ** 3
+        return  (10 ** self.precision) * ((10 ** 3 * (amount * self.totalSupply)) / self._totalAssets()) / 10 ** 3 / (10 ** self.precision)
```
Comment (fubuloubu) [inline contracts/Vault.vy:908]: So, the `10 ** 3` here was a precision fix. Remove these terms with your update

**Comment by banteg** [comment]
It would be a bit cleaner if we store the scaling factor instead of doing `10 ** self.precision` every time.
Something like `self.prec * ... / self.prec`

**Comment by steffenix** [comment]
> Overall, should find a way to test the boundaries of what is reasonable in the tokens we work with. Something like DAI is probably worst cade scenario, as it has 18 decimals and is ~$1 so we might see 9-10 digits of deposits (billions to tens of billions). Maybe go full worst case and do an analysis of what would happen if `totalDeposits` was `10 ** 30` (~$1t DAI)

For a token like dai the precision won't increase:

```
if self.decimals <= 18:
      self.precision = (18 - self.decimals)
    else:
      self.precision = 0
```

**Comment by fubuloubu** [comment]
> > Overall, should find a way to test the boundaries of what is reasonable in the tokens we work with. Something like DAI is probably worst cade scenario, as it has 18 decimals and is ~$1 so we might see 9-10 digits of deposits (billions to tens of billions). Maybe go full worst case and do an analysis of what would happen if `totalDeposits` was `10 ** 30` (~$1t DAI)
> 
> For a token like dai the precision won't increase:
> 
> ```
> if self.decimals <= 18:
>       self.precision = (18 - self.decimals)
>     else:
>       self.precision = 0
> ```

ah yes, so just make sure that numbers over 18 work too (the default being 0)

**Comment by steffenix** [comment]
> ah yes, so just make sure that numbers over 18 work too (the default being 0)

I have changed the comparison to be strictly under 18 so that it is tested.

File: contracts/Vault.vy [hunk 2]
```diff
@@ -791,7 +797,7 @@ def _issueSharesForAmount(to: address, amount: uint256) -> uint256:
     if totalSupply > 0:
         # Mint amount of shares based on what the Vault is managing overall
         # NOTE: if sqrt(token.totalSupply()) > 1e39, this could potentially revert
-        shares = amount * totalSupply / self._totalAssets()
+        shares = (self.precisionFactor) * amount * totalSupply / self._totalAssets() / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:800]:
```suggestion
        shares = self.precisionFactor * amount * totalSupply / self._totalAssets() / self.precisionFactor
```

File: contracts/Vault.vy [hunk 3]
```diff
@@ -884,9 +890,9 @@ def _shareValue(shares: uint256) -> uint256:
     freeFunds: uint256 = self._totalAssets()
 
     if(lockedFundsRatio < DEGREDATION_COEFFICIENT):
-        freeFunds -= (self.lockedProfit - (lockedFundsRatio * self.lockedProfit / DEGREDATION_COEFFICIENT))
+        freeFunds -= (self.lockedProfit - ((self.precisionFactor) * lockedFundsRatio * self.lockedProfit / DEGREDATION_COEFFICIENT/ (self.precisionFactor)))
```
Comment (fubuloubu) [inline contracts/Vault.vy:893]:
```suggestion
        freeFunds -= (
            self.lockedProfit
             - (
                 self.precisionFactor
                 * lockedFundsRatio
                 * self.lockedProfit
                 / DEGREDATION_COEFFICIENT
                 / self.precisionFactor
             )
         )
```

File: contracts/Vault.vy [hunk 4]
```diff
@@ -884,9 +890,9 @@ def _shareValue(shares: uint256) -> uint256:
...
     # NOTE: using 1e3 for extra precision here, when decimals is low
-    return ((10 ** 3 * (shares * freeFunds)) / self.totalSupply) / 10 ** 3
+    return (self.precisionFactor) * ((10 ** 3 * (shares * freeFunds)) / self.totalSupply) / 10 ** 3 / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:895]:
```suggestion
    return (
        self.precisionFactor
        * shares
        * freeFunds
        / self.totalSupply
        / self.precisionFactor
    )
```

File: contracts/Vault.vy [hunk 5]
```diff
@@ -896,7 +902,7 @@ def _sharesForAmount(amount: uint256) -> uint256:
     # See dev note on `deposit`.
     if self._totalAssets() > 0:
         # NOTE: if sqrt(token.totalSupply()) > 1e37, this could potentially revert
-        return ((10 ** 3 * (amount * self.totalSupply)) / self._totalAssets()) / 10 ** 3
+        return  (self.precisionFactor) * ((10 ** 3 * (amount * self.totalSupply)) / self._totalAssets()) / 10 ** 3 / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:905]:
```suggestion
        return  (
            self.precisionFactor
            * amount
            * self.totalSupply
            / self._totalAssets()
            / (self.precisionFactor
        )
```

File: contracts/Vault.vy [hunk 6]
```diff
@@ -1048,7 +1054,7 @@ def withdraw(
 
     # NOTE: This loss protection is put in place to revert if losses from
     #       withdrawing are more than what is considered acceptable.
-    assert totalLoss <= maxLoss * (value + totalLoss) / MAX_BPS
+    assert totalLoss <= (self.precisionFactor) * maxLoss * (value + totalLoss) / MAX_BPS / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1057]:
```suggestion
    assert totalLoss <= self.precisionFactor * maxLoss * (value + totalLoss) / MAX_BPS / self.precisionFactor
```

File: contracts/Vault.vy [hunk 7]
```diff
@@ -1381,7 +1387,7 @@ def removeStrategyFromQueue(strategy: address):
 @internal
 def _debtOutstanding(strategy: address) -> uint256:
     # See note on `debtOutstanding()`.
-    strategy_debtLimit: uint256 = self.strategies[strategy].debtRatio * self._totalAssets() / MAX_BPS
+    strategy_debtLimit: uint256 = (self.precisionFactor) * self.strategies[strategy].debtRatio * self._totalAssets() / MAX_BPS / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1390]:
```suggestion
    strategy_debtLimit: uint256 = (
        self.precisionFactor
        * self.strategies[strategy].debtRatio
        * self._totalAssets()
        / MAX_BPS
        / self.precisionFactor
    )
```

File: contracts/Vault.vy [hunk 8]
```diff
@@ -1413,9 +1419,9 @@ def _creditAvailable(strategy: address) -> uint256:
         return 0
 
     vault_totalAssets: uint256 = self._totalAssets()
-    vault_debtLimit: uint256 = self.debtRatio * vault_totalAssets / MAX_BPS
+    vault_debtLimit: uint256 = (self.precisionFactor) * self.debtRatio * vault_totalAssets / MAX_BPS / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1422]:
```suggestion
    vault_debtLimit: uint256 = self.precisionFactor * self.debtRatio * vault_totalAssets / MAX_BPS / self.precisionFactor
```

File: contracts/Vault.vy [hunk 9]
```diff
@@ -1413,9 +1419,9 @@ def _creditAvailable(strategy: address) -> uint256:
...
     vault_totalDebt: uint256 = self.totalDebt
-    strategy_debtLimit: uint256 = self.strategies[strategy].debtRatio * vault_totalAssets / MAX_BPS
+    strategy_debtLimit: uint256 = (self.precisionFactor) * self.strategies[strategy].debtRatio * vault_totalAssets / MAX_BPS / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1424]:
```suggestion
    strategy_debtLimit: uint256 = self.precisionFactor * self.strategies[strategy].debtRatio * vault_totalAssets / MAX_BPS / self.precisionFactor
```

File: contracts/Vault.vy [hunk 10]
```diff
@@ -1476,7 +1482,7 @@ def _expectedReturn(strategy: address) -> uint256:
     if timeSinceLastHarvest > 0 and totalHarvestTime > 0 and Strategy(strategy).isActive():
         # NOTE: Unlikely to throw unless strategy accumalates >1e68 returns
         # NOTE: Calculate average over period of time where harvests have occured in the past
-        return (self.strategies[strategy].totalGain * timeSinceLastHarvest) / totalHarvestTime
+        return (self.precisionFactor) * (self.strategies[strategy].totalGain * timeSinceLastHarvest) / totalHarvestTime / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1485]:
```suggestion
        return (
            self.precisionFactor
            * self.strategies[strategy].totalGain
            * timeSinceLastHarvest
            / totalHarvestTime
            / self.precisionFactor
        )
```

File: contracts/Vault.vy [hunk 11]
```diff
@@ -1517,7 +1523,7 @@ def _reportLoss(strategy: address, loss: uint256):
 
     # Also, make sure we reduce our trust with the strategy by the same amount
     debtRatio: uint256 = self.strategies[strategy].debtRatio
-    ratio_change: uint256 = min(loss * MAX_BPS / self._totalAssets(), debtRatio)
+    ratio_change: uint256 = min((self.precisionFactor) * loss * MAX_BPS / self._totalAssets() / (self.precisionFactor), debtRatio)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1526]:
```suggestion
    ratio_change: uint256 = min(self.precisionFactor * loss * MAX_BPS / self._totalAssets() / self.precisionFactor, debtRatio)
```

File: contracts/Vault.vy [hunk 12]
```diff
@@ -1527,13 +1533,15 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
     # NOTE: In effect, this reduces overall share price by the combined fee
     # NOTE: may throw if Vault.totalAssets() > 1e64, or not called for more than a year
     management_fee: uint256 = (
+        (self.precisionFactor) *
```
Comment (fubuloubu) [inline contracts/Vault.vy:1536]:
```suggestion
        self.precisionFactor *
```

File: contracts/Vault.vy [hunk 13]
```diff
@@ -1527,13 +1533,15 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
...
         (
             (self.strategies[strategy].totalDebt - Strategy(strategy).delegatedAssets())
             * (block.timestamp - self.strategies[strategy].lastReport)
             * self.managementFee
         )
         / MAX_BPS
         / SECS_PER_YEAR
+        / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1544]:
```suggestion
        / self.precisionFactor
```

File: contracts/Vault.vy [hunk 14]
```diff
@@ -1544,11 +1552,11 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
     # NOTE: No fee is taken when a Strategy is unwinding it's position, until all debt is paid
     if gain > 0:
         # NOTE: Unlikely to throw unless strategy reports >1e72 harvest profit
-        strategist_fee = (
+        strategist_fee = (self.precisionFactor) * (
```
Comment (fubuloubu) [inline contracts/Vault.vy:1555]:
```suggestion
        strategist_fee = (
            self.precisionFactor
```

File: contracts/Vault.vy [hunk 15]
```diff
@@ -1544,11 +1552,11 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
...
             gain * self.strategies[strategy].performanceFee
```
Comment (fubuloubu) [inline contracts/Vault.vy:1556]:
```suggestion
            * gain
            * self.strategies[strategy].performanceFee
```

File: contracts/Vault.vy [hunk 16]
```diff
@@ -1544,11 +1552,11 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
...
-        ) / MAX_BPS
+        ) / MAX_BPS / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1557]:
```suggestion
            / MAX_BPS
            / self.precisionFactor
        )
```

File: contracts/Vault.vy [hunk 17]
```diff
@@ -1544,11 +1552,11 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
...
         # NOTE: Unlikely to throw unless strategy reports >1e72 harvest profit
-        performance_fee = gain * self.performanceFee / MAX_BPS
+        performance_fee = (self.precisionFactor) * gain * self.performanceFee / MAX_BPS / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1559]:
```suggestion
        performance_fee = self.precisionFactor * gain * self.performanceFee / MAX_BPS / self.precisionFactor
```

File: contracts/Vault.vy [hunk 18]
```diff
@@ -1566,7 +1574,7 @@ def _assessFees(strategy: address, gain: uint256) -> uint256:
         # Send the rewards out as new shares in this Vault
         if strategist_fee > 0:  # NOTE: Guard against DIV/0 fault
             # NOTE: Unlikely to throw unless sqrt(reward) >>> 1e39
-            strategist_reward: uint256 = (strategist_fee * reward) / total_fee
+            strategist_reward: uint256 = (self.precisionFactor) * (strategist_fee * reward) / total_fee / (self.precisionFactor)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1577]:
```suggestion
            strategist_reward: uint256 = (
                self.precisionFactor
                * strategist_fee
                * reward
                / total_fee
                / self.precisionFactor
            )
```

**Review by fubuloubu — COMMENTED** [review]
In general, you can save a lot of gas if you load `self.precisionFactor` only once per function it's used in e.g.:
```python
    precisionFactor: uint256 = self.precisionFactor  # Caching to use only 1 SLOAD
```
then...
```python
    some_calc: uint256 = precisionFactor * ... / precisionFactor
```

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**